streaming/src/main/scala/org/apache/spark/streaming/scheduler/StreamingListener.scala

```diff
@@ -30,6 +30,12 @@ import org.apache.spark.util.Distribution
 sealed trait StreamingListenerEvent
 
 @DeveloperApi
+case class StreamingListenerBatchGenerateStarted(time: Long) extends StreamingListenerEvent
+
+@DeveloperApi
+case class StreamingListenerBatchGenerateCompleted(time: Long) extends StreamingListenerEvent
```

Comment:
you probably want the StreamingListenerBatchGenerateStarted & StreamingListenerBatchGenerateCompleted classes to have a BatchGenerateInfo class that has following fields & methods
   batchTime
   processingStartTime
   processingEndTime
   def processingDelay() : Long